stack-graphs/src/c.rs

```diff
@@ -614,6 +614,118 @@ pub extern "C" fn sg_stack_graph_add_edges(
     }
 }
 
+//-------------------------------------------------------------------------------------------------
+// Source info
+
+/// Contains information about a range of code in a source code file.
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+pub struct sg_source_info {
+    /// The location in its containing file of the source code that this node represents.
+    pub span: sg_span,
+    /// The kind of syntax entity this node represents (e.g. `function`, `class`, `method`, etc.).
+    pub syntax_type: sg_string_handle,
+    /// The full content of the line containing this node in its source file.
+    pub containing_line: sg_string_handle,
+}
+
+/// All of the position information that we have about a range of content in a source file
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+pub struct sg_span {
+    pub start: sg_position,
+    pub end: sg_position,
+}
+
+/// All of the position information that we have about a character in a source file
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+pub struct sg_position {
+    /// The 0-indexed line number containing the character
+    pub line: usize,
+    /// The offset of the character within its containing line
+    pub column: sg_offset,
+    /// The byte offset (within the file) of the line containing the character
+    pub containing_line: sg_bounds,
+    /// The byte offset (within the file) of the line containing the character, with any leading
+    /// and trailing whitespace removed
+    pub trimmed_line: sg_bounds,
+}
+
+/// The offset of a character within a string (typically a line of source code), using several
+/// different units
+///
+/// All offsets are 0-indexed.
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+pub struct sg_offset {
+    /// The number of UTF-8-encoded bytes appearing before this character in the string
+    pub utf8_offset: usize,
+    /// The number of UTF-16 code units appearing before this character in the string
+    pub utf16_offset: usize,
+}
+
+/// A half-open range identifying a range of characters in a string.
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+pub struct sg_bounds {
+    /// The index of the first character in the range.
+    pub start: usize,
+    /// The index of the first character _after_ the range.
+    pub end: usize,
+}
+
+/// An array of all of the source information in a stack graph.  Source information is associated
+/// with nodes, so node handles are indices into this array.  It is _not_ guaranteed that there
+/// will an entry in this array for every node handle; if you have a node handle whose value is
+/// larger than `count`, then use a 0-valued `sg_source_info` if you need source information for
+/// that node.
```

Comment:
Are users expected to index into this array themselves? This is an opportunity for errors (when `handle >= count`). If an accessor is provided, it can ensure the 0-valued `sg_source_info` for the user if necessary.

Reply:
We can add an accessor function, but I don't want to make it the mandatory way to access this data, because we're using this C API from Go.  Go has quite a lot of overhead for each C function call, and so it's important that we're providing a single array pointer that the Go code can index into directly.